File: server/game.py
import random
from server.config import PADDLE_HEIGHT, BALL_SPEED


class PongState:
    def __init__(self) -> None:
        self.ball_x = 0.5
        self.ball_y = 0.5
        self.ball_vx = BALL_SPEED
        self.ball_vy = BALL_SPEED * (random.random() - 0.5) * 2
        self.player_paddle_y = 0.5
        self.neural_paddle_y = 0.5
        self.player_score = 0
        self.neural_score = 0

    def update(self, neural_direction: float, player_y: float) -> None:
        self.player_paddle_y = max(
            PADDLE_HEIGHT / 2, min(1 - PADDLE_HEIGHT / 2, player_y)
        )
        self.neural_paddle_y = max(
            PADDLE_HEIGHT / 2,
            min(1 - PADDLE_HEIGHT / 2, self.neural_paddle_y + neural_direction),
        )

        self.ball_x += self.ball_vx
        self.ball_y += self.ball_vy

        # Wall bounce (top/bottom)
        if self.ball_y <= 0 or self.ball_y >= 1:
            self.ball_vy *= -1
            self.ball_y = max(0, min(1, self.ball_y))

        # Player paddle (left side, x ~ 0.05)
        if self.ball_x <= 0.07 and self.ball_vx < 0:
            if abs(self.ball_y - self.player_paddle_y) < PADDLE_HEIGHT / 2:
                self.ball_vx *= -1
                offset = (self.ball_y - self.player_paddle_y) / (PADDLE_HEIGHT / 2)
                self.ball_vy = offset * BALL_SPEED
            elif self.ball_x <= 0:
                self.neural_score += 1
                self._reset_ball(direction=1)

        # Neural paddle (right side, x ~ 0.95)
        if self.ball_x >= 0.93 and self.ball_vx > 0:
            if abs(self.ball_y - self.neural_paddle_y) < PADDLE_HEIGHT / 2:
                self.ball_vx *= -1
                offset = (self.ball_y - self.neural_paddle_y) / (PADDLE_HEIGHT / 2)
                self.ball_vy = offset * BALL_SPEED
            elif self.ball_x >= 1:
                self.player_score += 1
                self._reset_ball(direction=-1)

    def _reset_ball(self, direction: int) -> None:
        self.ball_x = 0.5
        self.ball_y = 0.5
        self.ball_vx = BALL_SPEED * direction
        self.ball_vy = BALL_SPEED * (random.random() - 0.5) * 2

    def to_dict(self) -> dict:
        return {
            "ball_x": self.ball_x,
            "ball_y": self.ball_y,
            "ball_vx": self.ball_vx,
            "ball_vy": self.ball_vy,
            "player_paddle_y": self.player_paddle_y,
            "neural_paddle_y": self.neural_paddle_y,
            "player_score": self.player_score,
            "neural_score": self.neural_score,
        }

File: web/app/engine/game.ts
import { BALL_SPEED, PADDLE_HEIGHT } from "./config";
import type { GameState } from "./types";

/**
 * Pong state machine. Both paddles are neural-driven in two-brain mode.
 */
export class PongGame {
  state: GameState;

  constructor() {
    this.state = {
      ballX: 0.5,
      ballY: 0.5,
      ballVx: BALL_SPEED,
      ballVy: BALL_SPEED * (Math.random() - 0.5) * 2,
      leftPaddleY: 0.5,
      rightPaddleY: 0.5,
      leftScore: 0,
      rightScore: 0,
    };
  }

  update(leftDirection: number, rightDirection: number): void {
    const s = this.state;
    const half = PADDLE_HEIGHT / 2;

    // Move paddles
    s.leftPaddleY = clamp(s.leftPaddleY + leftDirection, half, 1 - half);
    s.rightPaddleY = clamp(s.rightPaddleY + rightDirection, half, 1 - half);

    // Move ball
    s.ballX += s.ballVx;
    s.ballY += s.ballVy;

    // Wall bounce (top/bottom)
    if (s.ballY <= 0 || s.ballY >= 1) {
      s.ballVy *= -1;
      s.ballY = clamp(s.ballY, 0, 1);
    }

    // Left paddle (x ~ 0.05)
    if (s.ballX <= 0.07 && s.ballVx < 0) {
      if (Math.abs(s.ballY - s.leftPaddleY) < half) {
        s.ballVx *= -1;
        const offset = (s.ballY - s.leftPaddleY) / half;
        s.ballVy = offset * BALL_SPEED;
      } else if (s.ballX <= 0) {
        s.rightScore++;
        this.resetBall(1);
      }
    }

    // Right paddle (x ~ 0.95)
    if (s.ballX >= 0.93 && s.ballVx > 0) {
      if (Math.abs(s.ballY - s.rightPaddleY) < half) {
        s.ballVx *= -1;
        const offset = (s.ballY - s.rightPaddleY) / half;
        s.ballVy = offset * BALL_SPEED;
      } else if (s.ballX >= 1) {
        s.leftScore++;
        this.resetBall(-1);
      }
    }
  }

  private resetBall(direction: number): void {
    this.state.ballX = 0.5;
    this.state.ballY = 0.5;
    this.state.ballVx = BALL_SPEED * direction;
    this.state.ballVy = BALL_SPEED * (Math.random() - 0.5) * 2;
  }
}

function clamp(v: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, v));
}
